src/AWSCredentials.jl

```diff
@@ -56,392 +60,497 @@ mutable struct AWSCredentials
-    if force_refresh || will_expire(cr)
-        if debug_level > 0
-            println("Renewing credentials... ")
-        end
-        renew = cr.renew
-
-        if renew !== nothing
-            new_creds = renew()
-
-            new_creds === nothing && error("Can't find AWS credentials!")
-            copyto!(cr, new_creds)
-
-            # Ensure renewal function is not overwritten by the new credentials
-            cr.renew = renew
-        else
-            if debug_level > 0
-                println("Credentials cannot be renewed...")
-            end
-        end
-    end
-
-    return cr
-end
-
-function Base.show(io::IO,c::AWSCredentials)
-    println(io, string(c.user_arn,
-                       c.user_arn == "" ? "" : " ",
-                       "(",
-                       c.account_number,
-                       c.account_number == "" ? "" : ", ",
-                       c.access_key_id,
-                       c.secret_key == "" ? "" : ", $(c.secret_key[1:3])...",
-                       c.token == "" ? "" : ", $(c.token[1:3])..."),
-                       ")")
-end
-
-function Base.copyto!(dest::AWSCredentials, src::AWSCredentials)
-    for f in fieldnames(typeof(dest))
-        setfield!(dest, f, getfield(src, f))
-    end
-end
-import Base: copy!
-Base.@deprecate copy!(dest::AWSCredentials, src::AWSCredentials) copyto!(dest, src)
-
 
 """
-Is Julia running in an AWS Lambda sandbox?
-"""
-localhost_is_lambda() = haskey(ENV, "LAMBDA_TASK_ROOT")
+    localhost_is_ec2()
 
+Checks to see if the localhost is on EC2.
 
-"""
-Is Julia running on an EC2 virtual machine?
+# Returns
+- `Bool`: True if running the machine executing this is an EC2 instance
 """
 function localhost_is_ec2()
-
-    if localhost_is_lambda()
-        return false
+    # Checking to see if you are running on an EC2 instance is a complicated problem due to
+    # a large amount of caveats. Below is a list of methods to implement to work through
+    # most of these problems:
+    #
+    # 1. Check the `hostname -d`; this will not work if using non-Amazon DNS
+    # 2. Check metadata with EC2 internal domain name `curl -s
+    # http://instance-data.ec2.internal`; this will not work with a VPC (legacy EC2 only)
+    # 3. Check `sudo dmidecode -s bios-version`; this requires `dmidecode` on the instance
+    # 4. Check `/sys/devices/virtual/dmi/id/bios_version`; this may not work depending on
+    # the instance, Amazon does not document this file however so it's quite unreliable
+    # 5. Check `http://169.254.169.254`; This is a link-local address for metadata,
+    # apparently other cloud providers make this metadata URL available now as well so it's
+    # not guaranteed that you're on an EC2 instance
+    # 6. When checking the UUID, check for little-endian representation
+
+    # This is not guarenteed to work on Windows as RNG can make the UUID begin with EC2 on a
+    # non-EC2 instance
+    if @mock Sys.iswindows()
```

Comment:
Do we need to test this Windows logic on non-Windows systems?

Reply:
We do not have Windows machines on TravisCI, do you want those added in?

Reply:
I think you can leave this as is for now. I do find it a bit weird that we never actually test this on Windows though.